### PR TITLE
469 update retrofit dataset

### PR DESCRIPTION
**Pull request by longshuicy**
add additional function to update the existing inventory dataset with additional retrofit and mapping entry key information. https://github.com/IN-CORE/pyincore/pull/470/files#diff-d1b1cb98d647ccb4ebd9f558b2894472d298769fdd3c2bde8d9d4e00205e4de0R73

Since we use fiona object and looping through fiona is too slow, I used geopandas dataframe to merge and udpate; but then save to shapefile so we can resume using fiona to carry out normal fragility matching and calculation

----

### To test:
- Use this notebook to test SLC retrofit strategy
[SLC_retrofit_test.zip](https://github.com/IN-CORE/pyincore/files/14240182/SLC_retrofit_test.zip)
- Use this notebook to tests Galveston
[Galveston_retrofit_test.zip](https://github.com/IN-CORE/pyincore/files/14354234/Galveston_retrofit_test.zip)
- RUN `tests/pyincore/analyses/buildingdamage/test_buildingdamage_retrofit.py` and see results in the retrofit folder

Note:
- Joplin takes 31 seconds; 
- Lumberton takes 61 seconds.
- SLC if only retrofit 5000+ building with inner join total runtime: 158 seconds;
- Galveston retrofitting 1500 buildings combined damage takes 115 seconds

### CODE REVIEW

File: pyincore/dfr3service.py [hunk 1]
```diff
@@ -175,23 +177,32 @@ def create_dfr3_set(self, dfr3_set: dict, timeout=(30, 600), **kwargs):
         r = self.client.post(url, json=dfr3_set, timeout=timeout, **kwargs)
         return return_http_response(r).json()
 
-    def match_inventory(self, mapping: MappingSet, inventories: list, entry_key: str, add_info: list = None):
+    def match_inventory(self, mapping: MappingSet, inventories: list, entry_key: str = None, add_info: list = None):
```
Comment (Rashmil-1999) [inline pyincore/dfr3service.py:180]: Just a suggestion: you can add Optional[str] from typing here. Doesn't make a huge difference but it shows up in code suggestions when someone uses it.

File: pyincore/dfr3service.py [hunk 2]
```diff
@@ -255,21 +264,30 @@ def match_inventory(self, mapping: MappingSet, inventories: list, entry_key: str
 
         return dfr3_sets
 
-    def match_list_of_dicts(self, mapping: MappingSet, inventories: list, entry_key: str):
+    def match_list_of_dicts(self, mapping: MappingSet, inventories: list, entry_key: str = None):
```
Comment (longshuicy) [inline pyincore/dfr3service.py:267]: Optional[str]

**Comment by longshuicy** [comment]
Made one update to use inner join when joining with retrofit strategies and mapping; this will make sure only the building that has retrofit strategy will get computed again. Hence help with the performance.

**Comment by longshuicy** [comment]
Use this notebook to test SLC retrofit strategy
[SLC_retrofit_test.zip](https://github.com/IN-CORE/pyincore/files/14240182/SLC_retrofit_test.zip)

File: pyincore/utils/datasetutil.py [hunk 3]
```diff
@@ -63,3 +68,85 @@ def join_table_dataset_with_source_dataset(dataset, client):
...
+            inventory_df.rename(columns={
+                'retrofit_key': 'retrofit_k',
+                'retrofit_value': 'retrofit_v',
+                'description_mappingEntryKey': 'descr_map',
+                'config_mappingEntryKey': 'config_map'
+            }, inplace=True)
+
+            # save the updated inventory to a new shapefile
+            file_path = f"tmp_updated_{inventory_dataset.id}.shp"
+            inventory_df.to_file(file_path)
```
Comment (longshuicy) [inline pyincore/utils/datasetutil.py:146]: Can I use system temp folder on k8s datawolf? Use random string for the temp folder name

**Comment by ylyangtw** [comment]
Encountered FutureWarning in Galveston notebook:
```
/pyincore/utils/dataprocessutil.py:281: FutureWarning: The behavior of DataFrame.idxmax with all-NA values, or any-NA and skipna=False, is deprecated. In a future version this will raise ValueError
  max_key = dmg_result[dmg_states].idxmax(axis=1)
```
Just committed a fix if that's ok. Feel free to revert it if it affects anything.

Some test records from my end:
Joplin Tornado Retrofit execution time: 11.40898 seconds
Lumberton Flood Retrofit execution time: 44.84685 seconds
Galveston notebook runtime: 71 & 20
SLC notebook: 162 & 52

Good job on the hardwork!

**Review by ylyangtw — APPROVED** [review]

**Comment by longshuicy** [comment]
> Encountered FutureWarning in Galveston notebook:
> 
> ```
> /pyincore/utils/dataprocessutil.py:281: FutureWarning: The behavior of DataFrame.idxmax with all-NA values, or any-NA and skipna=False, is deprecated. In a future version this will raise ValueError
>   max_key = dmg_result[dmg_states].idxmax(axis=1)
> ```
> 
> Just committed a fix if that's ok. Feel free to revert it if it affects anything. Other than this, both notebooks and test work.
> 
> Some test records from my end: Joplin Tornado Retrofit execution time: 11.40898 seconds Lumberton Flood Retrofit execution time: 44.84685 seconds Galveston notebook runtime: 71 & 20 SLC notebook: 162 & 52
> 
> Good job on the hardwork!

Thanks for the fix and thorough test!

**Review by Vismayak — APPROVED** [review]

**Pull request merged**